packages/block-library/src/cover/edit/use-cover-is-dark.js

```diff
@@ -18,58 +18,66 @@ function retrieveFastAverageColor() {
 }
 
 /**
- * useCoverIsDark is a hook that returns a boolean variable specifying if the cover
- * background is dark or not.
+ * getCoverIsDark is a method that specifyies if the cover background is dark or not and
+ * applies the relevant attribute to help ensure that text is visible by default.
+ * This needs to be recalculated in all of the following Cover block scenarios:
+ * - When an overlay image is added, changed or removed
+ * - When the featured image is selected as the overlay image, or removed from the overlay
+ * - When the overlay color is changed
+ * - When the overlay color is removed
+ * - When the dimRatio is changed
  *
- * @param {?string} url          Url of the media background.
- * @param {?number} dimRatio     Transparency of the overlay color. If an image and
- *                               color are set, dimRatio is used to decide what is used
- *                               for background darkness checking purposes.
- * @param {?string} overlayColor String containing the overlay color value if one exists.
+ * See the comments below for more details about which aspects take priority when
+ * calculating the relative darkness of the Cover.
  *
- * @return {boolean} True if the cover background is considered "dark" and false otherwise.
+ * @return {Function} Function to calculate isDark attribute.
  */
-export default function useCoverIsDark( url, dimRatio = 50, overlayColor ) {
-	const [ isDark, setIsDark ] = useState( false );
-	useEffect( () => {
-		// If opacity is lower than 50 the dominant color is the image or video color,
-		// so use that color for the dark mode computation.
-		if ( url && dimRatio <= 50 ) {
-			const imgCrossOrigin = applyFilters(
-				'media.crossOrigin',
-				undefined,
-				url
-			);
-			retrieveFastAverageColor()
-				.getColorAsync( url, {
-					// Previously the default color was white, but that changed
-					// in v6.0.0 so it has to be manually set now.
-					defaultColor: [ 255, 255, 255, 255 ],
-					// Errors that come up don't reject the promise, so error
-					// logging has to be silenced with this option.
-					silent: process.env.NODE_ENV === 'production',
-					crossOrigin: imgCrossOrigin,
-				} )
-				.then( ( color ) => setIsDark( color.isDark ) );
-		}
-	}, [ url, url && dimRatio <= 50, setIsDark ] );
-	useEffect( () => {
-		// If opacity is greater than 50 the dominant color is the overlay color,
-		// so use that color for the dark mode computation.
-		if ( dimRatio > 50 || ! url ) {
-			if ( ! overlayColor ) {
-				// If no overlay color exists the overlay color is black (isDark )
-				setIsDark( true );
-				return;
+export default function useCoverIsDark() {
+	const getCoverIsDark = useCallback(
+		( url, dimRatio = 50, overlayColor ) => {
+			// If the dimRatio is less than 50, the image will have the most impact on darkness.
+			if ( url && dimRatio <= 50 ) {
```

Comment:
Before, when wrapped in an effect, this branch didn’t execute unless `url` had changed. Now, it may execute needlessly for changes to `overlayColor` or `dimRatio` (given `dimRatio` is less than `50`). Since the `url` hasn't changed the color average and resulting `isDark` determination will end up the same. Not sure it’s objectionable or worth optimizing but thought it should be noted.

Reply:
Thanks for pointing that out. I have done some testing and this doesn't seem to introduce any performance issues, even in the places where this might get called a lot, like dragging the mouse around the custom color picker, so given the extra complexity it would add in order to better track when and when not to run this, and the fact that is now only explicitly run in event handlers, I don't think it is worth trying to optimise it at this stage.